elyra-ai/airflow-notebook
# Add support for ELYRA_RUN_ID environment variable

This PR complements https://github.com/elyra-ai/elyra/pull/1732 by introducing a new protected environment variable `ELYRA_RUN_ID`, which is made available to notebooks | scripts during pipeline execution. The value is a unique string that changes for every pipeline run. For example
- run 1: `ELYRA_RUN_ID`=`some_value`
- run 2: `ELYRA_RUN_ID`=`some_other_value`

Notes:
- The actual value is the same for each node during pipeline execution.
- The actual value starts with the pipeline name, e.g. `my-pipeline-` followed by a sequence of characters that assure uniqueness.
- The trailing character sequence is generated by Apache Airflow at runtime and represents the run id.
- The actual value is not persisted by Elyra.
- If the user manually overrides the variable value in the node properties (e.g. sets `ELYRA_RUN_ID` to `bogusid`), the change is ignored, as it should be for all protected/system Elyra environment variables.
 
This PR also
 - fixes linting (using the same rules applied in `elyra-ai/elyra` and `elyra-ai/kfp-notebook`)
 - Makes lint a prerequisite to build tasks
 - Removes `flake8` as a runtime dependency

### How was this pull request tested?

- [Download/extract the test case archive](https://github.com/elyra-ai/kfp-notebook/files/6619921/1702_testcase.tar.gz) into a JL workspace
- Open `1702_test.pipeline` using the VPE (note that the comments indicate which nodes attempt to override the environment variable value
    ![image](https://user-images.githubusercontent.com/13068832/121258370-a9bd3900-c863-11eb-9e2f-a14653ef0620.png)
- Run the pipeline on AA for the first time. The output should indicate that the environment variable is set to `1702_test-MMDDhhmmss-scheduled...`. The first qualifier should be the DAG name,  which is the pipeline name concatenated with a date/timestamp. Example `ELYRA_RUN_ID` value: `1702_test5-0609143854-scheduled__2021-06-09T21:48:02.741530+00:00`
- Run the pipeline again (without making any changes) from the AA GUI. The output should now indicate that the environment variable is set to `1702_test-MMDDhhmmss-manual...`. The first qualifier should be the DAG name, which is the pipeline name concatenated with a date/timestamp. Example `ELYRA_RUN_ID` value: `1702_test5-0609143854-manual__2021-06-09T21:49:07.712340+00:00`
- Export the pipeline and review the generated Python script.


![image](https://user-images.githubusercontent.com/13068832/121436300-2ec45300-c935-11eb-980f-6eed526577df.png)


Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.
